QueryEditor: Adds inline query editor scene object

kaydelaney opened the pull request:
- Factored `PanelChromeRenderer` out of `VizPanelRenderer` to be used with  `QueryEditorRenderer`.
- Added generic utility function `getClosest` for finding the closest scene object that matches user-supplied criteria. Refactored existing sceneGraph functions to use this utility.

Leaving this PR as a draft for now in case there's any feedback or I've made any major mistakes 😅

torkelo reviewed (commented):
Great job!

torkelo commented:
Why would we use PanelChromeRenderer when rendering QueryEditor ?

dprokop commented on src/core/sceneGraph.ts:
```diff
@@ -6,85 +6,77 @@ import { SceneVariables } from '../variables/types';
+import { SceneQueryRunner } from '../querying/SceneQueryRunner';
+
+/** Walks up the scene graph, returning the first non-undefined result of `extract` */
+function getClosest<T>(sceneObject: SceneObject, extract: (s: SceneObject) => T | undefined): T | undefined {
+  let curSceneObject: SceneObject | undefined = sceneObject;
+  let extracted: T | undefined = undefined;
+
+  while (curSceneObject && !extracted) {
+    extracted = extract(curSceneObject);
+    curSceneObject = curSceneObject.parent;
+  }
+
+  return extracted;
+}
```
Nice!

dprokop commented on src/core/sceneGraph.ts:
```diff
@@ -6,85 +6,77 @@ import { SceneVariables } from '../variables/types';
-  }
-
-  if (scene.parent) {
-    return getLayout(scene.parent);
+  const layoutSceneObject = getClosest(scene, (s) => (isLayoutSceneObject(s) ? s : undefined));
+  if (layoutSceneObject) {
+    return layoutSceneObject;
   }
```
@kaydelaney - please sync with main, this was recently changed in https://github.com/grafana/scenes/pull/44

dprokop commented on src/components/PanelChromeRenderer.tsx:
```diff
@@ -0,0 +1,72 @@
+  statusMessage?: string;
+}
+
+export function PanelChromeRenderer({
```
@kaydelaney - not sure how this change is necessary for QueryEditor? Ref your comment in description: 

> tactored PanelChromeRenderer out of VizPanelRenderer to be used with QueryEditorRenderer.

dprokop commented on src/components/QueryEditor/QueryEditorRenderer.tsx:
```diff
@@ -0,0 +1,61 @@
+    model.setState({ query });
+
+    if (sceneQueryRunner) {
+      sceneQueryRunner.setState({
```
Technically we could start simple, and provide the `QueryEditor` scene object with a callback property for `onQueryChange` and allow the consumer to implement the wiring - similar to how we do in https://github.com/grafana/grafana/blob/main/public/app/features/scenes/scenes/demo.tsx#L121 

But I'm not sure about this. I think more and more patterns like this emerge making it super hard in the future to think about serializing scenes.

dprokop replied:
Thinking about this more and how potentially this could be used. Maybe the `QueryEditor` should have a local query state as it has right now, and an explicit **Run query** button rendered? Testing this now it appears that changes in the query editor are automatically propagated to `SceneQueryRunner`. While this is fine, and mimics the current Grafana behavior, I think It would be better and follow the changes we've been introducing to query editors in core (Run query button).

dprokop commented on src/components/QueryEditor/QueryEditorRenderer.tsx:
```diff
@@ -0,0 +1,61 @@
+
+  const QueryEditor = loadedDatasource.components.QueryEditor;
+
+  const onChange = (query: DataQuery) => {
```
@kaydelaney to follow some _yet undocumented_  best practices, this should belong to the model, not the renderer.

dprokop commented:
To summarize what we just discussed:
1. Let's change the implementation slightly so that it allows editing closest's SceneQueryRunner queries.
2. No add/remove button - basically one can only change what's defined in SQR.
3. Add demo in [scenes-app demos](https://github.com/grafana/scenes/tree/main/packages/scenes-app/src/pages/Demos).

kaydelaney commented:
I've implemented the changes that have been suggested and added a small demo.
One issue I've left unsolved for now is how to handle multiple queries (for now the editor component just deals with one query). Should we just have a scrollable list of query editors, or a tabbed view or something? What do you think @dprokop @torkelo

dprokop commented on packages/scenes/src/components/QueryEditor/QueryEditor.tsx:
```diff
@@ -0,0 +1,49 @@
+import { SceneQueryRunner } from '../../querying/SceneQueryRunner';
+
+export interface QueryEditorState extends SceneLayoutChildState {
+  datasource: string | DataSourceRef;
```
Given the suggested changes I think we should figure out data source based on the SceneQueryRunner rather than provide it explicitly. WDYT?

kaydelaney replied:
Ah, you're right, will fix

dprokop commented on packages/scenes-app/src/pages/Demos/scenes/queryEditor.tsx:
```diff
@@ -0,0 +1,25 @@
+          },
+        }),
+        new QueryEditor({
+          datasource: 'gdev-testdata',
```
FYI - regardless my other comment - we have a `const DATASOURCE_REF` defined that points to a datasource that is provisioned with the demo app.

dprokop reviewed (commented):
> One issue I've left unsolved for now is how to handle multiple queries (for now the editor component just deals with one query).

@kaydelaney I've just tested this, and also looking at the code multiple queries are supported. I'm confused :)

kaydelaney commented:
> > One issue I've left unsolved for now is how to handle multiple queries (for now the editor component just deals with one query).
> 
> @kaydelaney I've just tested this, and also looking at the code multiple queries are supported. I'm confused :)

Added support after I left that comment 😅

torkelo commented on packages/scenes/src/components/QueryEditor/QueryEditor.tsx:
```diff
@@ -0,0 +1,36 @@
+    super({ ...state });
+  }
+
+  public activate() {
+    super.activate();
+  }
```
can remove this

torkelo commented on packages/scenes/src/components/QueryEditor/QueryEditorRenderer.tsx:
```diff
@@ -0,0 +1,86 @@
+  const { data } = sceneGraph.getData(model).useState();
+  const sceneQueryRunner = sceneGraph.getSceneQueryRunner(model);
+
+  useEffect(() => {
+    const queryRunnerDatasource = sceneQueryRunner?.state.datasource;
+    getDataSource(queryRunnerDatasource, {})
+      .then((d) => model.setState({ datasource: d }))
+      .catch((err) => model.setState({ datasourceLoadErrorMessage: err }));
+  }, [sceneQueryRunner?.state.datasource, model]);
```
I would move this to the QueryEditor object class, and handle it inside an activation handler.

torkelo commented on packages/scenes/src/components/QueryEditor/QueryEditor.tsx:
```diff
@@ -0,0 +1,36 @@
+  datasourceLoadErrorMessage?: string;
+}
+
+export class QueryEditor extends SceneObjectBase<QueryEditorState> {
```
not sure what the best name for this is now that it handles multiple queries.

torkelo commented on packages/scenes/src/core/sceneGraph.ts:
```diff
@@ -6,69 +6,65 @@ import { VariableCustomFormatterFn, SceneVariables } from '../variables/types';
+/**
+ * Will walk up the scene object graph to the closest SceneQueryRunner
+ */
+export function getSceneQueryRunner(sceneObject: SceneObject): SceneQueryRunner | undefined {
+  return getClosest(sceneObject, (s) => (s.state.$data instanceof SceneQueryRunner ? s.state.$data : undefined));
+}
```
This would not find query runners that are wrapped in SceneDataTransformer.

torkelo commented on packages/scenes/src/components/QueryEditor/QueryEditorRenderer.tsx:
```diff
@@ -0,0 +1,86 @@
+  const { datasourceLoadErrorMessage, datasource } = model.useState();
+
+  const { data } = sceneGraph.getData(model).useState();
+  const sceneQueryRunner = sceneGraph.getSceneQueryRunner(model);
```
```suggestion
  const {queries} = sceneGraph.getSceneQueryRunner(model).useState();
```

torkelo reviewed (commented):
I feel pretty stupid for proposing this idea, It was much harder to make a generic/reusable scene object like this than I thought. 

Maybe we should move the whole QueryEditor component to scenes-app, and build a small Explore-like demo app there instead of adding this component to the scenes library. 

Until we have 1-2 clear use cases from plugins it feels like it could just cause confusion. 

thoughts @dprokop @kaydelaney ?

kaydelaney commented:
Implemented some of the suggested changes @torkelo

torkelo commented on packages/scenes/src/components/QueryEditor/QueryEditor.tsx:
```diff
@@ -0,0 +1,53 @@
+      queries: oldQueries.map((q) => (q.refId === query.refId ? query : q)),
+    });
+
+    sceneQueryRunner.runQueries();
```
not sure we should run queries on change automatically, only if onRunQuery is called.

dprokop replied:
Agree with that, think we should have explicit on button in the demo.

torkelo commented on packages/scenes/src/core/sceneGraph.ts:
```diff
@@ -4,69 +4,80 @@ import { DefaultTimeRange, EmptyDataNode, EmptyVariableSet } from '../variables/
+ * Will walk up the scene object graph to the closest SceneQueryRunner
+ */
+export function getSceneQueryRunner(sceneObject: SceneObject): SceneQueryRunner | undefined {
+  if (sceneObject.state.$data !== undefined) {
+    let curData: SceneDataProvider | undefined = sceneObject.state.$data;
+    while (curData !== undefined) {
+      if (curData instanceof SceneQueryRunner) {
+        return curData;
+      }
+
+      curData = curData.state.$data;
+    }
+  }
+  
+  if (sceneObject.parent) {
+    return getSceneQueryRunner(sceneObject.parent);
   }
 
-  throw new Error('No layout found in scene tree');
+  return undefined;
 }
```
I would like to wait with adding this to core lib until we know more of how scenes are constructed, and in what scenarios you need to find the closest query runner.

kaydelaney replied:
Sounds fine. Move to scenes-app package?

torkelo reviewed (commented):
Looks like a good base. I would move QueryEditor and QueryEditorRender 
to a scenes/explore-like demo scene for now where we can have this and over time add things to it maybe to test different scenarios and show how to do more advanced things

dprokop commented on packages/scenes/src/index.ts:
```diff
@@ -23,6 +23,8 @@ export { TextBoxVariable } from './variables/variants/TextBoxVariable';
 export { UrlSyncManager } from './services/UrlSyncManager';
 export { SceneObjectUrlSyncConfig } from './services/SceneObjectUrlSyncConfig';
 
+export { getDataSource } from './utils/getDataSource';
```
@kaydelaney let's not expose this util from the scenes ok? It technically has nothing to do with the API so I suggest just resolbing the datasource in the demo implementation.

dprokop reviewed (approved):
Looks good! Some final touches left to be made :)

dprokop commented:
@kaydelaney I've pushed sync with main and a minir layout fix

torkelo commented on packages/scenes-app/src/components/QueryEditor/QueryEditor.tsx:
```diff
@@ -0,0 +1,75 @@
+
+import { DataSourceApi } from '@grafana/data';
+
+export interface QueryEditorState extends SceneLayoutChildState {
```
```suggestion
export interface QueryEditorState extends SceneObjectState {
```

kaydelaney replied:
Gives an error: "An interface can only extend an object type or intersection of object types with statically known members"
